algopy_testing/src/algopy_testing/itxn.py

```diff
@@ -233,26 +411,91 @@ def __init__(
             **kwargs,
         }
-        self._parse_covariant_types()
+        self._parse_covariant_types(txn_type)
 
     def get_field(self, type_dict: object, name: str) -> typing.Any:
-        if name in type_dict.__annotations__:
+        if name in type_dict.__annotations__ or name in _TransactionFields.__annotations__:
             return self.fields.get(name)
 
         raise AttributeError(f"{type(self).__name__!r} object has no attribute {name!r}")
 
     def _parse_covariant_types(
         self,
+        txn_type: algopy.TransactionType,
     ) -> None:
         import algopy
 
+        # Define the covariant fields for each transaction type
+        covariant_fields = {
+            algopy.TransactionType.Payment: ["receiver", "close_remainder_to", "sender"],
+            algopy.TransactionType.KeyRegistration: [
+                "vote_key",
+                "selection_key",
+                "state_proof_key",
+                "sender",
+            ],
+            algopy.TransactionType.AssetConfig: [
+                "config_asset",
+                "unit_name",
+                "asset_name",
+                "url",
+                "metadata_hash",
+                "manager",
+                "reserve",
+                "freeze",
+                "clawback",
+                "sender",
+            ],
+            algopy.TransactionType.AssetTransfer: [
+                "xfer_asset",
+                "asset_sender",
+                "asset_receiver",
+                "asset_close_to",
+                "sender",
+            ],
+            algopy.TransactionType.AssetFreeze: ["freeze_asset", "freeze_account", "sender"],
+            algopy.TransactionType.ApplicationCall: [
+                "app_id",
+                "approval_program",
+                "clear_state_program",
+                "app_args",
+                "accounts",
+                "assets",
+                "apps",
+                "sender",
+            ],
+        }
+
+        # Get the relevant fields for the given transaction type
+        relevant_fields = covariant_fields.get(txn_type, [])
+
         for name, value in self.fields.items():
-            if isinstance(value, int):
-                self.fields[name] = algopy.UInt64(value)
-            if isinstance(value, bytes):
-                self.fields[name] = algopy.Bytes(value)
-            if isinstance(value, str):
-                self.fields[name] = algopy.Bytes(value.encode("utf-8"))
+            if name in relevant_fields:
+                if isinstance(value, int):
+                    self.fields[name] = algopy.UInt64(value)
+                elif isinstance(value, bytes):
+                    self.fields[name] = algopy.Bytes(value)
+                elif isinstance(value, str):
+                    self.fields[name] = algopy.Bytes(value.encode("utf-8"))
+                elif isinstance(value, tuple) and all(isinstance(v, bytes) for v in value):
```

Comment:
Should check for tuple first and then iterate each item then convert as tuples can contain a mixture of python and algopy types